### PR TITLE
 rgw/notification: Store the value of `persistent_queue` for existing topics and continue commiting events for all topics subscribed to given bucket

### PR DESCRIPTION
**Pull request by kchheda3**
Fixes https://tracker.ceph.com/issues/66097

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_pubsub.cc [hunk 1]
```diff
@@ -422,6 +422,11 @@ void RGWPSCreateTopicOp::execute(optional_yield y) {
                          << op_ret << dendl;
       return;
     }
+  } else if (already_persistent) {  // redundant call to CreateTopic
+    // initialize the persistent queue's location, using ':' as the namespace
+    // delimiter because its inclusion in a TopicName would break ARNs
+    dest.persistent_queue =
+        string_cat_reserve(get_account_or_tenant(s->owner.id), ":", topic_name);
```
Comment (cbodley) [inline src/rgw/rgw_rest_pubsub.cc:429]: i think this should just copy `dest.persistent_queue = topic->dest.persistent_queue`. the topic we're overwriting may have been a legacy one whose queue doesn't include the tenant namespace

Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:429]: would it not be better in way, the user now gets migrated to correct queue name ?
just that old queue will remain there in the system

Comment (cbodley) [reply src/rgw/rgw_rest_pubsub.cc:429]: not unless we have a way to clean up the old queue. the notification thread would keep polling on it forever

Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:429]: may be this code will still not make use of the new queue name even if it exist 
```    
if (struct_v >= 7) {
      decode(persistent_queue, bl);
    } else if (persistent) {
      // persistent topics created before v7 did not support tenant namespacing.
      // continue to use 'arn_topic' alone as the queue's rados object name
      persistent_queue = arn_topic;
    }

Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc]: > i think this should just copy `dest.persistent_queue = topic->dest.persistent_queue`. the topic we're overwriting may have been a legacy one whose queue doesn't include the tenant namespace

done

**Review by cbodley — APPROVED** [review]
looks correct to me. is there a test case we can extend with a re-create to verify that persistent topics still work?

**Comment by kchheda3** [comment]
> looks correct to me. is there a test case we can extend with a re-create to verify that persistent topics still work?

looked at `test_bn.py`, do not see any active test case that re-creates the topic, there is test case `test_ps_s3_topic_update` that updates the endpoint of the topic and does all re-verification but its disabled and also its purpose is more of testing update on endpoints and verifying the events are still received

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
> > looks correct to me. is there a test case we can extend with a re-create to verify that persistent topics still work?
> 
> looked at `test_bn.py`, do not see any active test case that re-creates the topic, there is test case `test_ps_s3_topic_update` that updates the endpoint of the topic and does all re-verification but its disabled and also its purpose is more of testing update on endpoints and verifying the events are still received and not about verifying the attributes. also currently get-topic does not return `persistent_queue` so not sure if adding new test case would verify this issue, unless we verify it via `radosgw-admin get topic`

can we add that to one of the existing persistent notification tests?
just create the topic twice. and verify that we still see the notifications.

**Comment by kchheda3** [comment]
> > > looks correct to me. is there a test case we can extend with a re-create to verify that persistent topics still work?
> > 
> > 
> > looked at `test_bn.py`, do not see any active test case that re-creates the topic, there is test case `test_ps_s3_topic_update` that updates the endpoint of the topic and does all re-verification but its disabled and also its purpose is more of testing update on endpoints and verifying the events are still received and not about verifying the attributes. also currently get-topic does not return `persistent_queue` so not sure if adding new test case would verify this issue, unless we verify it via `radosgw-admin get topic`
> 
> can we add that to one of the existing persistent notification tests? just create the topic twice. and verify that we still see the notifications.

updated the existing test case to re-create the topic and verify the persistent_queue is correctly populated.

**Review by yuvalif — APPROVED** [review]

**Comment by cbodley** [comment]
re-rerun looks good enough: https://pulpito.ceph.com/cbodley-2024-06-11_19:40:35-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**